src/Pact/Types/Exp.hs

```diff
@@ -155,50 +149,96 @@ litToPrim LDecimal {} = TyDecimal
-        , _eParsed :: !Parsed
-        } |
-  -- | Lists. '_eLitListType' distinguishes literal lists (`[1 2 3]`) from body forms.
-  EList { _eList :: ![Exp], _eLitListType :: !(Maybe (Type TypeName)), _eParsed :: !Parsed } |
-  -- | Object literals.
-  EObject { _eObject :: ![(Exp,Exp)], _eParsed :: !Parsed } |
-  -- | Special binding forms.
-  EBinding { _eBinding :: ![(Exp,Exp)], _eParsed :: !Parsed }
-           deriving (Eq,Generic)
-
-instance NFData Exp
-
-
-pattern EList' :: [Exp] -> Exp
-pattern EList' ls <- EList ls Nothing _
-pattern ELitList :: [Exp] -> Exp
-pattern ELitList ls <- EList ls (Just _) _
-pattern EAtom' :: Text -> Exp
-pattern EAtom' tag <- EAtom tag Nothing Nothing _
-pattern ELitString :: Text -> Exp
-pattern ELitString s <- ELiteral (LString s) _
-
-makePrisms ''Exp
-
+data ListDelimiter = Parens|Brackets|Braces deriving (Eq,Show,Ord,Generic,Bounded,Enum)
+instance NFData ListDelimiter
+
+listDelims :: ListDelimiter -> (Text,Text)
+listDelims Parens = ("(",")")
+listDelims Brackets = ("[","]")
+listDelims Braces = ("{","}")
+
+enlist :: ListDelimiter -> ((Text,Text) -> a) -> a
+enlist d f = f (listDelims d)
+
+data Separator = Colon|ColonEquals|Comma deriving (Eq,Ord,Generic,Bounded,Enum)
+instance NFData Separator
+instance Show Separator where
+  show Colon = ":"
+  show ColonEquals = ":="
+  show Comma = ","
+
+class ExpInfo b a | a -> b where
+  eInfo :: a -> b
+
+instance ExpInfo Info Info where
+  eInfo = id
+
+data LiteralExp i = LiteralExp
+  { _litLiteral :: !Literal
+  , _litInfo :: !i
+  } deriving (Eq,Ord,Generic,Functor,Foldable,Traversable)
+instance Show (LiteralExp i) where show LiteralExp{..} = show _litLiteral
+instance ExpInfo i (LiteralExp i) where
+  eInfo = _litInfo
+instance NFData i => NFData (LiteralExp i)
+
+data AtomExp i = AtomExp
+  { _atomAtom :: !Text
+  , _atomQualifiers :: ![Text]
+  , _atomInfo :: i
+  } deriving (Eq,Ord,Generic,Functor,Foldable,Traversable)
+instance Show (AtomExp i) where
+  show AtomExp{..} = intercalate "." (map unpack $ _atomQualifiers ++ [_atomAtom])
+instance ExpInfo i (AtomExp i) where
+  eInfo = _atomInfo
+instance NFData i => NFData (AtomExp i)
+
+data ListExp i = ListExp
+  { _listList :: ![(Exp i)]
+  , _listDelimiter :: !ListDelimiter
+  , _listInfo :: !i
+  } deriving (Eq,Ord,Generic,Functor,Foldable,Traversable)
+instance Show (ListExp i) where
+  show ListExp{..} =
+    enlist _listDelimiter $ \(o,c) ->
+      unpack o ++ unwords (map show _listList) ++ unpack c
+instance ExpInfo i (ListExp i) where
+  eInfo = _listInfo
+instance NFData i => NFData (ListExp i)
+
+data SeparatorExp i = SeparatorExp
+  { _sepSeparator :: !Separator
+  , _sepInfo :: !i
+  } deriving (Eq,Ord,Generic,Functor,Foldable,Traversable)
+instance Show (SeparatorExp i) where show (SeparatorExp{..}) = show _sepSeparator
+instance ExpInfo i (SeparatorExp i) where
+  eInfo = _sepInfo
+instance NFData i => NFData (SeparatorExp i)
+
+
+-- | Pact syntax expressions
+data Exp i =
```

Comment:
I think it's fair to call all of these constructors expressions, except for `ESeparator`. It seems to me that separators are a different sort.

I was playing with similar ideas recently and had a slightly different factoring, maybe worth considering.

https://github.com/kadena-io/pact/blob/399d9bd8048ed60a5aa3d0e73e2eae6c31ad7a0e/src/Pact/Types/SExp.hs#L48-L58

Reply:
I agree with @joelburget here - we need to nail down what is and is not a valid expression/term/token when we can all meet next.

Reply:
I realized that in my representation you could also have `Token (Punctuation ":" TrailingSpace) : SExp`, so it has the same problem if you consider it that.